File: backend/src/main/java/com/jordan/chat_system/service/MessageService.java
package com.jordan.chat_system.service;

import com.jordan.chat_system.dto.EditMessageRequest;
import com.jordan.chat_system.dto.MessageResponse;
import com.jordan.chat_system.dto.SendMessageRequest;
import com.jordan.chat_system.entity.Message;
import org.springframework.data.domain.Page;

import java.util.List;

public interface MessageService {

    MessageResponse sendMessage(
            String senderEmail,
            SendMessageRequest request
    );

    Page<MessageResponse> getConversation(Long senderId, Long receiverId, int page, int size);

    List<Message> markMessagesAsDelivered(String receiverEmail);

    List<Message> markMessagesAsRead(Long senderId, Long receiverId);

    long countUnreadMessages(
            Long senderId,
            Long receiverId
    );

    Message deleteMessage(
            Long messageId,
            String currentUserEmail
    );

    Message editMessage(
            Long messageId,
            String currentUserEmail,
            EditMessageRequest request
    );

    List<MessageResponse> searchConversation(
            Long userId,
            Long contactId,
            String query
    );
}

File: frontend/src/services/messageService.js
import api from "../api/axios";

export async function getConversation(receiverId) {
  const response = await api.get(`/messages/conversation/${receiverId}`);

  return response.data;
}

export async function sendMessage(request) {
  const response = await api.post("/messages", request);

  return response.data;
}

export async function markAsRead(senderId) {
  await api.post(`/messages/read/${senderId}`);
}

export async function deleteMessage(messageId) {
  const response = await api.delete(`/messages/${messageId}`);
  return response.data;
}

export async function editMessage(messageId, content) {
  const response = await api.patch(`/messages/${messageId}`, {
    content
  });

  return response.data;
}

export async function searchConversation(contactId, query) {
  const response = await api.get("/messages/search", {
    params: {
      contactId,
      query
    },    
  });

  return response.data;
}
